src/parseTools.js

```diff
@@ -1023,3 +1035,7 @@ function getPerformanceNow() {
     return 'performance.now';
   }
 }
+
+function implicitSelf() {
+  return ENVIRONMENT.includes('node') ? 'self.' : '';
```

Comment:
What happens if the environment includes both node and web and we run the result on the web?    i.e. isn't choosing to use `self.` or not here at compile time too early.. don't we need to decide at runtime maybe?

Reply:
In the web environment, it can run with or without the `self.` prefix, but in the node environment, it must have the `self.` prefix to run.
So I think the above code is correct.

Reply:
Ah I see.  It looks like we have the same pattern in `src/worker.js` but we simply always use `self` there.

I wonder if we could find a way to avoid the `self.` completely by using `defineProperty` on the global object under node?  Maybe that would be more code overall?

Reply:
I think this issue is the same as the [issue](https://github.com/emscripten-core/emscripten/pull/20188#discussion_r1338809434) below, it can create a separate PR after merging this to deal with these two issues together.